contracts/Market.sol

```diff
@@ -941,144 +772,72 @@ contract Market is
    * to transfer their ERC20 tokens.
    * @param amount The total purchase amount in ERC20 tokens. This is the total number of removals being
    * purchased, scaled by the price multiple.
-   * @param customFee The custom fee percentage that was paid to Nori, as an integer, specified here for
-   * inclusion in emitted events.
-   * @param customPriceMultiple The custom price that will be charged for this transaction.
+   * @param customFee The fee percentage that was paid to Nori off chain, as an integer, specified here
+   * for inclusion in emitted events.
+   * @param customPriceMultiple The price that will be charged for this transaction.
+   * @param supplier The only supplier address from which to purchase carbon removals in this transaction, or
+   * zero address if any supplier is valid.
+   * @param vintages The valid set of vintages from which to fulfill this order, empty if any vintage is valid.
    */
   function swapWithoutFeeSpecialOrder(
     address recipient,
     address purchaser,
     uint256 amount,
     uint256 customFee,
-    uint256 customPriceMultiple
+    uint256 customPriceMultiple,
+    address supplier,
+    uint256[] memory vintages
```

Comment:
I don't understand

Reply:
If I'm understanding right, there's a special order purchase function and a non special order purchase function. Is it possible to also merge these two into just one? I realize the special order one uses this modifier to prevent unauthorized access for zero fee purchases, but perhaps we can validate/handle/determine that by using an if/else inside a single swap function

Does that make sense?

Reply:
oh yeah that makes sense... interesting -- like basically just policing it based on whether the sender has the required role instead of using a modifier on the signature itself which gives us the same permission but more granular control.

the issue I see with this is that the signature of this mega-function would have to look like the special order purchase function signature, with all these extra parameters that have special meanings when they have special values such as the zero address or empty array (not to mention the complexity of specifying your price and fee!).  The complicated signature works for us because we know how to provide an intentional value for each field, but I don't believe the public-facing swap endpoint should have most of these parameters.

maybe one way to do it could be to refactor so that there's one internal swap function that basically looks like the current special order case, and then the public facing function wraps that and provides the arguments to the internal function:

```
...
customFee: getNoriFeePercentage();
priceMultiple: getPriceMultiple();
supplier: address(0);
vintages: new uint256[](0);
})
```

Reply:
and the you also have a special order wrapper that has all of the parameters.  that still leaves 2 different functions though and honestly doesn't do much to simplify the guts and would increase gas costs for the public endpoint

Reply:
Yeah basically what you're saying is what I was thinking (although, alternatively, I don't see why the public facing endpoint couldnt just require zero values as well since they're easy enough to specify-- especially if this at some point gets further simplified to handle things off chain using removal ID specification somehow instead as all you'd need to zero out is a single array).

Using your suggestion might still simplify the readability/audit ability as the two entry points simply pipe into the same internal function, and do nothing else otherwise other than set default values.

Reply:
Likewise you could also combine the two swap functions in a similar way using either approach by inferring the path based on whether the value was a zero value or not.

But you might be right when you say we might not gain much if all we end up doing is create several if/else forks later on. Though maybe that's addressable with well written cases with enums (e.g., pseudocode: `checkoutType =  sender === admin ? Swap.SpecialOrder : v,r,s === 0 ? Swap.WithoutPermit : ... ` followed by if statements for each swap enums value... 🤷Super rough idea, not suggesting we do it now, but worth considering when thinking about where all this seems to be heading)